test/test_coverart_image.py

```diff
@@ -126,6 +145,483 @@ def test_set_data(self):
+            },
+            'expected': {
+                1: {},
+                2: {'cover.png': '1'},
+                3: {'cover.png': '2'},
+            },
+            'types': {
+                '1': ['front'],
+                '2': ['back'],
+            }
+        }
+        self._save_images(cases)
+
+    def test_save_types_2(self):
+        cases = {
+            'options': {
+                'caa_image_type_as_filename': True,
+                'cover_image_filename': 'cover',
+                'save_images_overwrite': False,
+            },
+            'expected': {
+                1: {},
+                2: {'cover.png': '1'},
+                3: {'back.png': '2', 'cover.png': '1'},
+            },
+            'types': {
+                '1': ['front'],
+                '2': ['back'],
+            }
+        }
+        self._save_images(cases)
+
+    def test_save_types_overwrite_2(self):
+        cases = {
+            'options': {
+                'caa_image_type_as_filename': True,
+                'cover_image_filename': 'cover',
+                'save_images_overwrite': True,
+            },
+            'expected': {
+                1: {},
+                2: {'cover.png': '1'},
+                3: {'back.png': '2', 'cover.png': '1'},
+            },
+            'types': {
+                '1': ['front'],
+                '2': ['back'],
+            }
+        }
+        self._save_images(cases)
+
+    def test_save_types_3(self):
+        cases = {
+            'options': {
+                'caa_image_type_as_filename': True,
+                'cover_image_filename': 'cover',
+                'save_images_overwrite': False,
+            },
+            'expected': {
+                1: {},
+                2: {'cover.png': '1'},
+                3: {'cover (1).png': '2', 'cover.png': '1'}
+            },
+            'types': {
+                '1': ['front'],
+                '2': ['front'],
+            }
+        }
+        self._save_images(cases)
+
+    def test_save_types_overwrite_3(self):
+        cases = {
+            'options': {
+                'caa_image_type_as_filename': True,
+                'cover_image_filename': 'cover',
+                'save_images_overwrite': True,
+            },
+            'expected': {
+                1: {},
+                2: {'cover.png': '1'},
+                3: {'cover.png': '2'}
+            },
+            'types': {
+                '1': ['front'],
+                '2': ['front'],
+            }
+        }
+        self._save_images(cases)
+
+    def test_save_types_4(self):
+        cases = {
+            'options': {
+                'caa_image_type_as_filename': True,
+                'cover_image_filename': 'cover',
+                'save_images_overwrite': False,
+            },
+            'expected': {
+                1: {},
+                2: {'front.png': '1'},
+                3: {'cover.png': '2', 'front.png': '1'},  # FIXME: is this what we expect?
```

Comment:
I guess this is expected if explicitly marking an image not as `is_front`. But this should be a theoretical case for now (unless we allow users to modify the type, which could happen). But this is another example why I find `caa_image_type_as_filename` problematic.

Reply:
Yes, `caa_image_type_as_filename` is problematic.

Reply:
After recent discussion that we also had for the cover art options refactoring this would indeed be expected behaviour. If `image_type_as_filename` is set it will use the type as filename for all images except for the cover. The cover will use the `cover_image_filename`.

This example is intentionally confusing I guess, but the image with type "back" is marked as "Is_front", while the image with type "front" is not. Hence the back type image will get the "cover" name and the front type image gets named by type, which is "front".

In practice this could happen only with a badly behaving cover art provider. Current providers would mark type "front" as "is_front".